api/v1beta2/imagepolicy_types.go

```diff
@@ -0,0 +1,174 @@
+type ImagePolicySpec struct {
+	// ImageRepositoryRef points at the object specifying the image
+	// being scanned
+	// +required
+	ImageRepositoryRef meta.NamespacedObjectReference `json:"imageRepositoryRef"`
+	// Policy gives the particulars of the policy to be followed in
+	// selecting the most recent image
+	// +required
+	Policy ImagePolicyChoice `json:"policy"`
+	// FilterTags enables filtering for only a subset of tags based on a set of
+	// rules. If no rules are provided, all the tags from the repository will be
+	// ordered and compared.
+	// +optional
+	FilterTags *TagFilter `json:"filterTags,omitempty"`
+}
+
+// ImagePolicyChoice is a union of all the types of policy that can be
+// supplied.
+type ImagePolicyChoice struct {
+	// SemVer gives a semantic version range to check against the tags
+	// available.
+	// +optional
+	SemVer *SemVerPolicy `json:"semver,omitempty"`
+	// Alphabetical set of rules to use for alphabetical ordering of the tags.
+	// +optional
+	Alphabetical *AlphabeticalPolicy `json:"alphabetical,omitempty"`
+	// Numerical set of rules to use for numerical ordering of the tags.
+	// +optional
+	Numerical *NumericalPolicy `json:"numerical,omitempty"`
+}
+
+// SemVerPolicy specifies a semantic version policy.
+type SemVerPolicy struct {
+	// Range gives a semver range for the image tag; the highest
+	// version within the range that's a tag yields the latest image.
+	// +required
+	Range string `json:"range"`
+}
+
+// AlphabeticalPolicy specifies a alphabetical ordering policy.
+type AlphabeticalPolicy struct {
+	// Order specifies the sorting order of the tags. Given the letters of the
+	// alphabet as tags, ascending order would select Z, and descending order
+	// would select A.
+	// +kubebuilder:default:="asc"
+	// +kubebuilder:validation:Enum=asc;desc
+	// +optional
+	Order string `json:"order,omitempty"`
+}
+
+// NumericalPolicy specifies a numerical ordering policy.
+type NumericalPolicy struct {
+	// Order specifies the sorting order of the tags. Given the integer values
+	// from 0 to 9 as tags, ascending order would select 9, and descending order
+	// would select 0.
+	// +kubebuilder:default:="asc"
+	// +kubebuilder:validation:Enum=asc;desc
+	// +optional
+	Order string `json:"order,omitempty"`
+}
+
+// TagFilter enables filtering tags based on a set of defined rules
+type TagFilter struct {
+	// Pattern specifies a regular expression pattern used to filter for image
+	// tags.
+	// +optional
+	Pattern string `json:"pattern"`
+	// Extract allows a capture group to be extracted from the specified regular
+	// expression pattern, useful before tag evaluation.
+	// +optional
+	Extract string `json:"extract"`
+	// Discriminator allows to split the tags in multiple groups and execute the
+	// policy against each of those groups. This value if present is a capture
+	// group to be extracted from the regular expression pattern.
+	// +optional
+	Discriminator string `json:"discriminator,omitempty"`
```

Comment:
Thing is I do not want to use the setter policy, I want to have ephemeral environment, so if I have 3 open PR, I want to have 3 environments.  Setter policy is just a setter on yaml file, it cannot help me there as it would create a single environment.
If I want to create 3 environments I would have to create 3 policies with associated kubernetes objects but then I feel it contradict the gitops approach as the CI would have to create deployment objet (and then why use flux to automate what is already done by the CI) 

It is why there is another PR on the automation controller to be able to use one policy and duplicate some objects.

The goal would be to to be able to create a docker image on the CI (and delete some when PR is closed)  and then flux on its own would be able to trigger what need to be triggered.

Reply:
In general, changes to code can be accompanied by changes to configuration. For example, if you add a feature to your app, you might have some new argument that needs to be supplied to the image; or, to have a feature flag set, to enable the feature. For this reason, I think of preview environments as being driven by changes to configs (YAMLs), rather than new images; i.e., automation belongs outside the rather restricted view of the image reflector.